### ISSUE TITLE
Memory Heap issue with ning instrumentation

### ISSUE THREAD

**Issue by twcrone** [newrelic/newrelic-java-agent#783]
A Java app that runs with multiple JVM allocations in PROD is consistently reaching max memory heap usage, causing the instances to fail.  It appears that some of our `ning` instrumentation might be the culprit.

https://newrelic.atlassian.net/browse/GTSE-13769
https://newrelic.atlassian.net/browse/GTSE-14104

(Migrate to Jira)

**Comment by jasonjkeller** [comment]
We suspect that this was introduced in agent 7.5.0 either due to changes to the [Caffeine caching library](https://github.com/newrelic/newrelic-java-agent/pull/593) used by the agent or some changes to the [ning-async-http-client instrumentation](https://github.com/newrelic/newrelic-java-agent/pull/513).

For now the issue can be worked around by disabling the `ning-async-http-client` instrumentation by adding the following under the pre-existing `class_transformer` stanza in the agent yaml:

```
  class_transformer:
    com.newrelic.instrumentation.ning-async-http-client-1.0:
      enabled: false
    com.newrelic.instrumentation.ning-async-http-client-1.1:
      enabled: false
    com.newrelic.instrumentation.ning-async-http-client-1.6.1:
      enabled: false
```

**Comment by kford-newrelic** [comment]
Not selected for the Jan-Mar quarter. Will consider for a future quarter.

**Comment by jasonjkeller** [comment]
Related comment:

> Based on the memory dumps I am seeing, the "responseStatus" attribute of the ning AsyncHolder is being retained post completion of the segment. The fix for the issue I linked earlier was the set this property to null on completion and on throwables updates. I believe, however, that this property is being set after these callback functions are called and the segment is nulled out. (I don't have a way to confirm this, it's just a hypothesis based on what I'm seeing inside of the heap dump)
>
> From what I can also see from the heap dump, the majority of items stuck in memory seem to be from keep-alive connections.

The referenced change in https://github.com/newrelic/newrelic-java-agent/pull/1672 went into the 8.9.0 release.

**Comment by jasonjkeller** [comment]
This issues should be fixed in version 8.12.0 of the agent by https://github.com/newrelic/newrelic-java-agent/pull/1903
